src/OpenDB/src/lefout/lefout.cpp

```diff
@@ -1014,3 +1063,62 @@ bool lefout::writeTechAndLib(dbLib* lib, const char* lef_file)
 
   return true;
 }
+
+bool lefout::writeAbstractLef(dbBlock* db_block, const char* lef_file)
+{
+  _out = fopen(lef_file, "w");
+
+  if (_out == nullptr) {
+    logger_->error(utl::ODB, 1072, "Cannot open LEF file %s\n", lef_file);
+    return false;
+  }
+
+  writeHeader(db_block);
+  writeBlock(db_block);
+
+//  for (dbNet* net :db_block->getNets()) {
+//    dbWireShapeItr wire_shape_itr;
+//
+//    dbShape shape;
+//    for (wire_shape_itr.begin(net->getWire()); wire_shape_itr.next(shape);) {
+//
+//    }
+//  }
+  fprintf(_out, "END LIBRARY\n");
+
+
+  return true;
+}
+
+void lefout::writeBlock(dbBlock* db_block)
+{
+  dbBox* bounding_box = db_block->getBBox();
+  double origin_x = lefdist(bounding_box->xMin());
+  double origin_y = lefdist(bounding_box->xMin());
+  double size_x = lefdist(bounding_box->getDX());
+  double size_y = lefdist(bounding_box->getDY());
+
+  fprintf(_out, "MACRO %s\n", db_block->getName().c_str());
+  fprintf(_out, "  FOREIGN %s 0 0 ;\n", db_block->getName().c_str());
+  fprintf(_out, "  CLASS BLOCK ;\n");
+  fprintf(_out, "  ORIGIN %g %g ;\n", origin_x, origin_y);
+  fprintf(_out, "  SIZE %g %g ;\n", size_x, size_y);
+  writePins(db_block);
+  fprintf(_out, "END %s\n", db_block->getName().c_str());
+  fclose(_out);
+}
+void lefout::writePins(dbBlock* db_block)
+{
+  for (dbBTerm* b_term : db_block->getBTerms()) {
+    fprintf(_out, "  PIN %s\n", b_term->getName().c_str());
+    fprintf(_out, "    DIRECTION %s ;\n", b_term->getIoType().getString());
+    fprintf(_out, "    USE %s ;\n", b_term->getSigType().getString());
+    for (dbBPin* db_b_pin : b_term->getBPins()) {
+       fprintf(_out, "    PORT\n");
+       dbSet<dbBox> term_pins = db_b_pin->getBoxes();
+       writeBoxes(&term_pins, "      ");
```

Comment:
Why not pass by const & rather than * (avoid the need for a null check)

Reply:
Changed this to by reference, but the begin method on the iterator of dbSet isn't const so I cannot make it const.